modules/basicfuncs/list-funcs.c

```diff
@@ -328,3 +329,305 @@ tf_implode(LogMessage *msg, gint argc, GString *argv[], GString *result)
 }
 
 TEMPLATE_FUNCTION_SIMPLE(tf_implode);
+
+typedef enum _StringMatchMode
+{
+  SMM_LITERAL = 0,
+  SMM_PREFIX,
+  SMM_SUBSTRING,
+  SMM_GLOB,
+  SMM_PCRE
+} StringMatchMode;
+
+typedef struct _StringMatcher
+{
+  StringMatchMode mode;
+  gchar *pattern;
+  GPatternSpec *glob;
+  pcre *pcre;
+  pcre_extra *pcre_extra;
+} StringMatcher;
+
```

Comment:
I think we should move this to lib instead of keeping them in this module. It pretty much has everything to be a module on its own right. Later we can work on merging the common pieces between this and the logmatcher module.

Reply:
Actually I already have a working-branch doing this with a bit of cleanup. I would not like to include it in this PR, as it is old enough, and was reviewed many times. The refactor will come soon :)

The refactored API looks like this currently (changes might be needed, but I think this should be in a different PR's scope):
```
typedef struct _StringMatcher StringMatcher;

gboolean string_matcher_prepare(StringMatcher *self, gpointer user_data);
gboolean string_matcher_match(StringMatcher *self, const char *string, gsize string_len);
void string_matcher_free(StringMatcher *self);

StringMatcher *string_matcher_literal_new(const gchar *pattern);
StringMatcher *string_matcher_prefix_new(const gchar *pattern);
StringMatcher *string_matcher_substring_new(const gchar *pattern);
StringMatcher *string_matcher_glob_new(const gchar *pattern);
StringMatcher *string_matcher_pcre_new(const gchar *pattern);
```